### PR TITLE
Fix row slicing with sort_and_finalize throw on column slicing when the column group is larger than the segment column size

### LINKED ISSUE #1848: Gracefully fail when appending in either sort and finalize or plain finalize if existing data is column sliced
- [ ] Include test of the reverse, that writing wide non-column sliced data and then appending directly (i.e. not via incompletes) works and is readable

### LINKED ISSUE #1851: Add sort and finalize stress test
Use large data (i.e. not relying on config_context  to meddle with defaults). If it takes too long to run in general we don't have to merge it, although with lmdb should be OK

### PR DESCRIPTION
**Pull request by vasil-pashov**
#### What does this implement or fix?
Closes #1851 
Closes #1848 
#### Checklist

<details>
  <summary>
   Checklist for code changes...
  </summary>
 
 - [ ] Have you updated the relevant docstrings, documentation and copyright notice?
 - [ ] Is this contribution tested against [all ArcticDB's features](../docs/mkdocs/docs/technical/contributing.md)?
 - [ ] Do all exceptions introduced raise appropriate [error messages](https://docs.arcticdb.io/error_messages/)?
 - [ ] Are API changes highlighted in the PR description?
 - [ ] Is the PR labelled as enhancement or bug so it appears in autogenerated release notes?
</details>

### CODE REVIEW

File: python/tests/unit/arcticdb/version_store/test_sort_merge.py [hunk 1]
```diff
@@ -612,4 +612,37 @@ def test_update_symbol_list(lmdb_library):
     lib.write(sym_2, df, staged=True)
     lib.sort_and_finalize_staged_data(sym_2, mode=StagedDataFinalizeMethod.APPEND)
     assert lib_tool.count_keys(KeyType.SYMBOL_LIST) == 3
-    assert set(lib.list_symbols()) == set([sym, sym_2])
\ No newline at end of file
+    assert set(lib.list_symbols()) == set([sym, sym_2])
+
+class TestSlicing:
```
Comment (vasil-pashov) [inline python/tests/unit/arcticdb/version_store/test_sort_merge.py:617]: Set the config `Merge.SegmentSize` to something lower than the default `100000` rows to make the tests faster.

File: cpp/arcticdb/version/version_core.cpp [hunk 2]
```diff
@@ -1456,6 +1456,14 @@ VersionedItem sort_merge_impl(
         "Finalizing staged data is not allowed with empty staging area"
     );
 
+    user_input::check<ErrorCode::E_INVALID_USER_ARGUMENT>(
+        write_options.dynamic_schema || pipeline_context->staged_descriptor_->field_count() < write_options.column_group_size,
+        "Sorting and finalizing staged data is not implemented in the case when column slicing would appear. The "
+        "input DataFrame has {} fields which is more than the column group size ({}) set in the library options",
+        pipeline_context->staged_descriptor_->field_count(),
+        write_options.column_group_size
+    );
```
Comment (alexowens90) [inline cpp/arcticdb/version/version_core.cpp:1465]: This shouldn't be an error? It should just write segments wider than the lib config setting

File: cpp/arcticdb/version/version_core.cpp [hunk 3]
```diff
@@ -1456,6 +1456,14 @@ VersionedItem sort_merge_impl(
         "Finalizing staged data is not allowed with empty staging area"
     );
 
+    user_input::check<ErrorCode::E_INVALID_USER_ARGUMENT>(
+        write_options.dynamic_schema || pipeline_context->staged_descriptor_->field_count() < write_options.column_group_size,
+        "Sorting and finalizing staged data is not implemented in the case when column slicing would appear. The "
```
Comment (DrNickClarke) [inline cpp/arcticdb/version/version_core.cpp:1461]: not yet implemented?

we do have a plan to solve this in future

File: python/tests/unit/arcticdb/version_store/test_sort_merge.py [hunk 4]
```diff
@@ -612,4 +614,41 @@ def test_update_symbol_list(lmdb_library):
...
+    def test_long_append_segment(self, lmdb_library):
+        set_config_int('Merge.SegmentSize', 5)
```
Comment (alexowens90) [inline python/tests/unit/arcticdb/version_store/test_sort_merge.py:621]: Use a context manager to ensure it is reset at the end of the test:
https://github.com/man-group/ArcticDB/blob/008fa54cfe123c9ac4276643ebb5e22126d7a3c0/python/arcticdb/util/test.py#L122

File: python/tests/unit/arcticdb/version_store/test_sort_merge.py [hunk 5]
```diff
@@ -612,4 +614,41 @@ def test_update_symbol_list(lmdb_library):
     lib.write(sym_2, df, staged=True)
     lib.sort_and_finalize_staged_data(sym_2, mode=StagedDataFinalizeMethod.APPEND)
     assert lib_tool.count_keys(KeyType.SYMBOL_LIST) == 3
-    assert set(lib.list_symbols()) == set([sym, sym_2])
\ No newline at end of file
+    assert set(lib.list_symbols()) == set([sym, sym_2])
+
+class TestSlicing:
```
Comment (alexowens90) [inline python/tests/unit/arcticdb/version_store/test_sort_merge.py:619]: Can we also have a test where the individual staged segments are smaller than the segment row count, but when they are compacted they are larger?

**Review by alexowens90 — CHANGES_REQUESTED** [review]

File: cpp/arcticdb/version/version_core.cpp (hunk 3 above)
Comment (vasil-pashov) [reply cpp/arcticdb/version/version_core.cpp]: In the long term, yes. It would require a large refactor though.

**Review by alexowens90 — APPROVED** [review]

**Pull request merged**